libraries/AC_Avoidance/AC_Avoid.cpp

```diff
@@ -261,15 +300,20 @@ void AC_Avoid::adjust_roll_pitch(float &roll, float &pitch, float veh_angle_max)
  *
  * Uses velocity adjustment idea from Randy's second email on this thread:
  * https://groups.google.com/forum/#!searchin/drones-discuss/obstacle/drones-discuss/QwUXz__WuqY/qo3G8iTLSJAJ
+ * acceleration is limited to zero when limiting is active, this removes unpredictable (for the pilot) feed forward effects
  */
-void AC_Avoid::limit_velocity(float kP, float accel_cmss, Vector2f &desired_vel_cms, const Vector2f& limit_direction, float limit_distance_cm, float dt)
+void AC_Avoid::limit_accel_velocity(float kP, float accel_cmss, Vector2f &desired_vel_cms, Vector2f &desired_accel_cms, const Vector2f& limit_direction, float limit_distance_cm, float dt)
 {
     const float max_speed = get_max_speed(kP, accel_cmss, limit_distance_cm, dt);
     // project onto limit direction
     const float speed = desired_vel_cms * limit_direction;
     if (speed > max_speed) {
         // subtract difference between desired speed and maximum acceptable speed
         desired_vel_cms += limit_direction*(max_speed - speed);
+
+        // also limit acceleration in direction
+        desired_accel_cms.zero();
```

Comment:
I wonder what affect this has on the vehicle in the direction parallel to the wall.  For example, if the vehicle reaches the marginat an angle.  Does it instantly lose it's forward desired acceleration?  I wonder if this makes it jerk or get stuck to the wall instead of sliding.

Reply:
I tested trying a bunch of clever things with the acceleration and I couldn't get it to do anything better. For example we currently limit the velocity to a given acceleration value but don't bother setting the defied acceleration to that value, as far as I can tell it makes no difference if you do it properly or not. 

The only thing this was added for is to remove the 'spring back'. If your holding full stick towards a obstacle but the velo is limited to zero when you let go the copter assumes it needs to decelerate so its leans back. Now it does nothing. I think is nice for the pilot, the spring back is not intuitive.

https://youtu.be/LllCOrNLWVc?t=65

vs 

https://youtu.be/LllCOrNLWVc?t=118

Reply:
I totally agree that the spring back is not good.  It perhaps doesn't matter but it didn't always spring back.  I'm sure that crept in when Leonard rewrote the Loiter controller so the pilot had direct control of the desired acceleration.

I'd really like to see a test of what happens when the vehicle is sliding along the wall... and the moment that it first reaches the margin.

Reply:
I think instead of setting the desired_accel_cms to zero it needs to be limited so that it doesn't project towards the margin just like the velocity vector.

Reply:
The spring back is the accleration vector projecting away from the limit. You can fly round in loiter just fine with it zeroed all the time.

Reply:
It's hard to believe that simply zeroing the desired acceleration is the optimal solution.  Perhaps we need input from @lthall.  I'm pretty sure he wrote the comment, "// TODO: We need to also limit the _desired_accel" in AC_Loiter.cpp.